src/datasets/utils/tf_utils.py

```diff
@@ -68,3 +72,335 @@ def is_numeric_feature(feature):
+            "columns_to_ranks": self.columns_to_ranks,
+        }
+        for i in range(num_workers):
+            worker_name = f"datasets_tf_worker_{i}"
+            names.append(worker_name)
+
+            worker_shape_shms = {
+                col: SharedMemory(name=f"{worker_name}_{col}_shape", create=True, size=rank * 8)
+                for col, rank in self.columns_to_ranks.items()
+            }
+            worker_shape_arrays = {
+                col: np.ndarray(shape=(shape_shm.size // 8,), dtype=np.int64, buffer=shape_shm.buf)
+                for col, shape_shm in worker_shape_shms.items()
+            }
+            shape_shms.append(worker_shape_shms)
+            shape_arrays.append(worker_shape_arrays)
+
+            worker_indices = per_worker_batches[i]
+            if i == final_batch_worker and final_batch is not None:
+                final_batch_arg = final_batch
+            else:
+                final_batch_arg = None
+            worker_kwargs = {
+                "worker_name": worker_name,
+                "indices": worker_indices,
+                "extra_batch": final_batch_arg,
+                "array_ready_event": array_ready_events[i],
+                "array_loaded_event": array_loaded_events[i],
+                **base_args,
+            }
+            worker = ctx.Process(target=self.worker_loop, kwargs=worker_kwargs, daemon=True)
+            worker.start()
+            workers.append(worker)
+
+        end_signal_received = False
+        while not end_signal_received:
+            for i in range(num_workers):
+                array_ready_events[i].wait()
+                array_ready_events[i].clear()
+                array_shapes = shape_arrays[i]
+                array_sizes = {
+                    col: np.prod(shape) * np.dtype(self.columns_to_np_types[col]).itemsize
+                    for col, shape in array_shapes.items()
+                }
+                if any(size < 0 for size in array_sizes.values()):
+                    end_signal_received = True
+                    break
+                array_shms = {
+                    col: SharedMemory(name=f"{names[i]}_{col}", create=False, size=size)
+                    for col, size in array_sizes.items()
+                }
+                arrays = {
+                    col: np.copy(
+                        np.ndarray(shape=array_shapes[col], dtype=self.columns_to_np_types[col], buffer=shm.buf)
+                    )
+                    for col, shm in array_shms.items()
+                }
+                yield arrays
+                for shm in array_shms.values():
+                    shm.close()
+                array_loaded_events[i].set()
+        # Now we just do some cleanup
+        for worker, shape_shms in zip(workers, shape_shms):
+            worker.join()
+            for shm in shape_shms.values():
+                shm.unlink()
+
+    def __call__(self):
+        return self
+
+    @staticmethod
+    def worker_loop(
```

Comment:
some logic here is shared with `dataset_to_tf`, which could perhaps be shared :p